pyprof/nvtx/nvmarker.py

```diff
@@ -274,18 +276,69 @@ def patchClass(cls):
         if isfunc(cls, f):
             add_wrapper(cls, f)
 
-
-def init():
-    print("Initializing NVTX monkey patches")
+# Monkey-patch all classes in torch
+#
+def patch_torch_classes():
     for cls in [
             torch,
             torch.Tensor,
             torch.nn.functional,
     ]:
         patchClass(cls)
 
+
+# Monkey-patch all forward functions in torch.nn libraries
+#
+def patch_torch_nn_forward_functions():
     for cls in [torch.nn.RNN, torch.nn.RNNCell, torch.nn.LSTM, torch.nn.LSTMCell, torch.nn.GRU, torch.nn.GRUCell]:
         if isfunc(cls, 'forward'):
             add_wrapper(cls, 'forward')
 
+
+# Monkey-patch the dataloader in torch.utils.data
+#
+def patch_dataloader():
+    mod = torch.utils.data.dataloader
+    old_iter = mod.DataLoader.__iter__
+
+    def new_iter(self, *args, **kwargs):
+
+        # Push trace marker
+        nvtx.range_push(traceMarker())
+
+        # First pass is for creating the dataloader + returning the first data
+        cadena = argMarker(mod, "DataLoader", args, kwargs)
+        nvtx.range_push(cadena)
+
+        for x in old_iter(self, *args, **kwargs):
+
+            # Pop tracemarker
+            nvtx.range_pop();           
+
+            # Dataloader stop, Model start
+            nvtx.range_pop() 
+
+            yield x
+
+            # Push trace marker
+            nvtx.range_push(traceMarker())
+
+            # Model stop, dataloader start
+            cadena = argMarker(mod, "DataLoader", args, kwargs)
+            nvtx.range_push(cadena)
+
+        # Pop the last iteration before returning
+        nvtx.range_pop()
+        nvtx.range_pop()
+
+    mod.DataLoader.__iter__ = new_iter
```

Comment:
Can this be moved to under ```old_iter = mod.DataLoader.__iter__```?
I think that would read better.

Alternatively, I would move the inner function up to the first defined part of the ```patch_dataloader``` function.

Reply:
I don't think I can.  I can't reference new_iter until after it is defined, and since the function references old_iter it has to be after that.  I think the function has to keep the same prototype, so I can't pass in old_iter in order to move its location in the file.

Reply:
Yea i think you are right. Python looks to be 1 pass. 
Resolving.